workflows/eval: Request reviews from changed package maintainers

infinisil opened the pull request:
Request reviews from maintainers of changed packages, mirroring what ofborg does, but without having to wait for it anymore. Part of https://github.com/NixOS/nixpkgs/issues/355847

## Things done

- [x] Tested: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/37#event-15696464782

---

This work is funded by [Tweag](https://tweag.io) and [Antithesis](https://antithesis.com/) :sparkles:

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
workflows/eval: Request reviews from changed package maintainers

Currently we need to rely on ofborg requesting reviews from package
maintainers, which takes a while with ofborg's eval queue. Since
recently we're doing faster evaluations with GitHub Actions, which contain all
necessary information to determine reviewers of changed packages the
same way ofborg does. This PR takes advantage of that.

TheGiraffe3 commented on .github/workflows/eval.yml:
```diff
@@ -217,6 +224,14 @@ jobs:
     steps:
+      # See ./codeowners-v2.yml, reuse the same App because we need the same permissions
+      # Can't use the token received from permissions above, because it can't get enough permissions
+      - uses: actions/create-github-app-token@5d869da34e18e7287c1daad50e0b8ea0f506ce69 # v1.11.0
```
Why not just use v1.11.0 instead of a commit hash if they are the same thing?

Mic92 replied:
It's a security feature. Dependabot supports updating these hashes natively.

FliegendeWurst commented on ci/eval/compare/maintainers.nix:
```diff
@@ -0,0 +1,123 @@
+    builtins.map (
+      pkg:
+      builtins.map (maintainer: {
+        handle = lib.toLower maintainer.github;
```
If possible I would use the `githubId` instead, as sometimes accounts are renamed and the list not updated. (see #362419)

infinisil replied:
Done!

GaetanLepage commented on ci/eval/compare/utils.nix:
```diff
@@ -52,12 +56,12 @@ rec {
-        { name = "hello"; platform = "x86_64-darwin"; }
-        { name = "bye"; platform = "aarch64-darwin"; }
-        { name = "bye"; platform = "x86_64-darwin"; }
+        { name = "hello"; platform = "aarch64-linux"; packagePath = [ ... ]; }
```
This also needs to be changed in the documentation of `groupByPlatform` and `groupByKernel`.
Maybe you can just add `...` instead:
`{ name = "hello"; platform = "aarch64-linux"; }` --> `{ name = "hello"; platform = "aarch64-linux"; ... }`

infinisil committed:
workflows/eval: Use maintainer GitHub IDs for review requests of changed packages

The handles can change over time and there's nothing guaranteeing the
ones in the maintainer list are up-to-date. In comparison GitHub IDs
never change.

infinisil commented:
Still works: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/37#event-15709044331

JohnRTitor reviewed (commented):
Eval "Process" seems to be failing.

infinisil commented:
@JohnRTitor A bit unfortunate but unfixable right now, because the workflow for this PR itself uses the version from the base branch, but fetches the code from the PR, which now takes an extra argument not passed in the workflow

Mic92 commented:
> @JohnRTitor A bit unfortunate but unfixable here, because the workflow for this PR itself uses the version from the base branch, but fetches the code from the PR, which now takes an extra argument not passed in the workflow

This is the reference to look at: https://github.com/Infinisil-s-Test-Organization/nixpkgs/pull/37#event-15696464782

FliegendeWurst commented on .github/workflows/eval.yml:
```diff
@@ -251,6 +266,20 @@ jobs:
+
+          # maintainers.json contains GitHub IDs. Look up handles to request reviews from.
+          # There appears to be no API to request reviews based on GitHub IDs
+          jq -r 'keys[]' comparison/maintainers.json \
+            | while read -r id; do gh api /user/"$id"; done \
+            | jq -s '{ reviewers: map(.login) }' \
+            > reviewers.json
```
```suggestion
          jq -r 'keys[]' comparison/maintainers.json \
            | head -n15 \
            | while read -r id; do gh api /user/"$id"; done \
            | jq -s '{ reviewers: map(.login) }' \
            > reviewers.json
```
GitHub apps are limited to 15 reviewer requests anyway, and this way treewide changes are handled better. Adding a `head -n15` here should work, but I did not test.

Mic92 replied:
Sounds like a good idea. Should be the list sorted, so it's always the same 15 reviewers?

Mic92 replied:
I think there is also a limit with posting maintainers in comments.

FliegendeWurst commented on .github/workflows/eval.yml:
```diff
@@ -217,6 +224,14 @@ jobs:
       pull-requests: write
       statuses: write
     steps:
+      # See ./codeowners-v2.yml, reuse the same App because we need the same permissions
+      # Can't use the token received from permissions above, because it can't get enough permissions
```
The limit only applies per run, but I'm not sure. And codeowners is another actions run.

JohnRTitor commented on ci/eval/compare/maintainers.nix:
```diff
@@ -0,0 +1,123 @@
+      builtins.trace "Package has no maintainers: ${pkg.name}" false
+  ) attrsWithMaintainers;
+
+  relevantFilenames =
+    drv:
+    (lib.lists.unique (
+      builtins.map (pos: lib.strings.removePrefix (toString ../..) pos.file) (
+        builtins.filter (x: x != null) [
+          (builtins.unsafeGetAttrPos "maintainers" (drv.meta or { }))
+          (builtins.unsafeGetAttrPos "src" drv)
+          # broken because name is always set by stdenv:
+          #    # A hack to make `nix-env -qa` and `nix search` ignore broken packages.
+          #    # TODO(@oxij): remove this assert when something like NixOS/nix#1771 gets merged into nix.
+          #    name = assert validity.handled; name + lib.optionalString
+          #(builtins.unsafeGetAttrPos "name" drv)
+          (builtins.unsafeGetAttrPos "pname" drv)
+          (builtins.unsafeGetAttrPos "version" drv)
+
+          # Use ".meta.position" for cases when most of the package is
+          # defined in a "common" section and the only place where
+          # reference to the file with a derivation the "pos"
+          # attribute.
+          #
+          # ".meta.position" has the following form:
+          #   "pkgs/tools/package-management/nix/default.nix:155"
+          # We transform it to the following:
+          #   { file = "pkgs/tools/package-management/nix/default.nix"; }
+          { file = lib.head (lib.splitString ":" (drv.meta.position or "")); }
+        ]
+      )
+    ));
```
@infinisil

Mic92 replied:
I feel like we can first start with a minimal feature set and make it more complex later. Now since ofborg has been shutdown and will be re-instantiated with only a smaller feature set.

JohnRTitor reviewed (approved):
Okay LGTM, let's just merge this initial PR, we can improve later based on feedbacks recieved

FliegendeWurst commented:
It seems there are some limitations for who we can request reviews from.

> gh: Reviews may only be requested from collaborators. One or more of the users or teams you specified is not a collaborator of the NixOS/nixpkgs repository. (HTTP 422)

https://github.com/NixOS/nixpkgs/actions/runs/12580696560/job/35063365466

Mic92 commented:
@FliegendeWurst https://github.com/NixOS/nixpkgs/pull/370186

FliegendeWurst commented:
That only applies for the other issue https://github.com/NixOS/nixpkgs/issues/355847#issuecomment-2567511203

Mic92 commented:
Ah right, we need a special github token than.

Mic92 commented:
We should probably revert this pull request until these two issues are addressed.

FliegendeWurst commented:
> Ah right, we need a special github token than.

I don't think it's possible to request reviews from non-members. So a better token would not help.

JohnRTitor commented:
> I don't think it's possible to request reviews from non-members

Exactly, you either need to be a repo collaborator or an org member.

symphorien commented:
A different failure mode in https://github.com/NixOS/nixpkgs/actions/runs/12588895670/job/35088361244
```

Run # Get all currently set rebuild labels
gh: Review cannot be requested from pull request author. (HTTP 422)
{"message":"Review cannot be requested from pull request author.","documentation_url":"https://docs.github.com/rest/pulls/review-requests#request-reviewers-for-a-pull-request","status":"422"}
```

emilazy commented:
> Yes, we might have failing CI jobs, but ultimately we don't lose anything: The only thing that is not happening with those kind of failures is the request for maintainer reviews.

I get an email for every single PR I open telling me the job failed, and every PR has a big red ❌ on it that makes it look like eval failed. I think those are sufficient cause to revert this until it can be fixed.

GGG-KILLER commented:
> I get an email for every single PR I open telling me the job failed, and every PR has a big red ❌ on it that makes it look like eval failed. I think those are sufficient cause to revert this until it can be fixed.

Instead of reversing this change, can't we temporarily add a `||:` to the line that adds the reviewers?
That way it'll keep requesting them in the cases it works and not block PRs when it fails.

emilazy commented:
Sure, that also works if someone PRs it. I just think we shouldn’t be too hesitant to revert PRs that cause treewide issues, since we didn’t have this PR 16 hours ago anyway and it is cheap to merge again with fixes but leaving things merged in a broken state multiplies the problems they cause. Better to revert and bring it back without the issue than to leave the problem present while waiting for a fix.

GGG-KILLER commented:
> Sure, that also works if someone PRs it. I just think we shouldn’t be too hesitant to revert PRs that cause treewide issues, since we didn’t have this PR 16 hours ago anyway and it is cheap to merge again with fixes but leaving things merged in a broken state multiplies the problems they cause. Better to revert and bring it back without the issue than to leave the problem present while waiting for a fix.

I'm not against reversing this, I'd just prefer to have a somewhat working reviewer request automation instead of having to wait 2 business months for ofborg to request them itself, but either work for me.

I can't PR the hack fix right now but I'm willing to do it, if someone wants to revert it meanwhile I'm okay with it as well, better than having a CI check giving a false negative on a PR.

infinisil commented:
I'm kind of on holidays, but I can make a quick fix, working on it..

Atemu commented:
We can also just revert until you're back.

infinisil commented:
Maybe that would've been better, but here's the fix, already tested: https://github.com/NixOS/nixpkgs/pull/370456 :)
